src/solc_0.8/assetERC721/BaseERC721.sol

```diff
@@ -0,0 +1,150 @@
+        require(from == _msgSender() || isApprovedForAll(from, _msgSender()), "!AUTHORIZED");
+        _burn(id);
+    }
+
+    /// @notice Transfer token with given id.
+    /// @dev Required by IMintableERC721.
+    /// @param from Address whose token is to be transferred.
+    /// @param to Recipient.
+    /// @param tokenId The token id to be transferred.
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 tokenId
+    ) public override(ERC721Upgradeable, IMintableERC721, IERC721Token) {
+        ERC721Upgradeable.safeTransferFrom(from, to, tokenId);
+    }
+
+    /// @notice Query if a token id exists.
+    /// @param tokenId Token id to be queried.
+    function exists(uint256 tokenId) external view override(IMintableERC721, IERC721Token) returns (bool) {
+        return _exists(tokenId);
+    }
+
+    /// @notice Query if a contract implements interface `id`.
+    /// @param id the interface identifier, as specified in ERC-165.
+    /// @return `true` if the contract implements `id`.
+    function supportsInterface(bytes4 id)
+        public
+        view
+        override(AccessControlUpgradeable, ERC721Upgradeable)
+        returns (bool)
+    {
+        return super.supportsInterface(id);
+    }
+
+    /// @notice Change the address of the trusted forwarder for meta-transactions
+    /// @param trustedForwarder The new trustedForwarder
+    function setTrustedForwarder(address trustedForwarder) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _trustedForwarder = trustedForwarder;
+    }
+
+    /// @notice Change the base uri for token metadata
+    /// @param baseUri The new base uri.
+    function setBaseUri(string calldata baseUri) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        baseTokenURI = baseUri;
+    }
+
+    function isTrustedForwarder(address forwarder) public view returns (bool) {
+        return forwarder == _trustedForwarder;
+    }
+
+    function getTrustedForwarder() external view returns (address trustedForwarder) {
+        return _trustedForwarder;
+    }
+
+    function _msgSender() internal view virtual override returns (address sender) {
+        if (isTrustedForwarder(msg.sender)) {
+            // The assembly code is more direct than the Solidity version using `abi.decode`.
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                sender := shr(96, calldataload(sub(calldatasize(), 20)))
+            }
+        } else {
+            return msg.sender;
+        }
+    }
+
+    function _msgData() internal view virtual override returns (bytes calldata) {
+        if (isTrustedForwarder(msg.sender)) {
+            return msg.data[:msg.data.length - 20];
+        } else {
+            return msg.data;
+        }
+    }
+
+    function _baseURI() internal view override returns (string memory) {
+        return baseTokenURI;
+    }
+
+    uint256[50] private __gap2; // In case
```

Comment:
2 gaps ?

Reply:
Going to look more into this and talk to Andres